### PR TITLE
Migrate ProvidesMethodFactoryCodeGen to KSP

### PR DESCRIPTION
**Pull request by ZacSweers**
Ref #751.

### CODE REVIEW

**Comment by ZacSweers** [comment]
@JoelWilcox @RBusarow this is ready now 👍. Last major one of the simple generators.

File: compiler/build.gradle [hunk 1]
```diff
@@ -37,6 +37,7 @@ dependencies {
 
   compileOnly libs.auto.service.annotations
   compileOnly libs.kotlin.compiler
+  compileOnly libs.ksp.compilerPlugin
```
Comment (ZacSweers) [inline compiler/build.gradle:40]: This is for internal API access, see comment below

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/ProvidesMethodFactoryCodeGen.kt [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+                property.getter?.isAnnotationPresent<Provides>() == true
+            }
+            .map { property ->
+              CallableReference.from(property)
+            }
+
+          val className = clazz.toClassName()
+          val containingFile = clazz.containingFile!!
+          // TODO we need a public API for this in KSP
+          //  https://github.com/google/ksp/issues/1621
```
Comment (ZacSweers) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/ProvidesMethodFactoryCodeGen.kt:114]: They're open to this, so I'll send a PR to them soon. In the meantime, we do this internal access in #713 anyway so it's not necessarily new

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/ProvidesMethodFactoryGenerator.kt
Comment (JoelWilcox) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/ProvidesMethodFactoryGenerator.kt:1]: Could you add a commit just for the file/class rename? To help highlight what is changed/unchanged, and potentially preserve some of the history depending on if we squash later

Comment (ZacSweers) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/ProvidesMethodFactoryGenerator.kt:1]: I'd have to restart the PR from scratch and I'm not sure if you can keep that history given the squash and merge?

Comment (JoelWilcox) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/ProvidesMethodFactoryGenerator.kt:1]: Not sure what you mean by starting from scratch, you should be able to just squash your changes, commit the file rename, and then commit the rest. And then with just those two commits there's no need to squash and merge 😁 . Next time though! No worries for this one

Comment (ZacSweers) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/ProvidesMethodFactoryGenerator.kt:1]: ahh I thought you meant while preserving the other commits (which I did try to keep somewhat focused). 👍

File: compiler/src/test/java/com/squareup/anvil/compiler/dagger/ProvidesMethodFactoryGeneratorTest.kt [hunk 3]
```diff
@@ -2556,6 +2558,8 @@ public final class DaggerComponentInterface implements ComponentInterface {
       }
       """,
     ) {
+      // KSP always resolves the inferred return type
+      assumeFalse(mode is AnvilCompilationMode.Ksp)
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply compiler/src/test/java/com/squareup/anvil/compiler/dagger/ProvidesMethodFactoryGeneratorTest.kt]: 303dd4ff193fb6c96e3a7db8d415338ee8ce0a81

[the remark this answers is not in the document]
Comment (JoelWilcox) [reply compiler/src/test/java/com/squareup/anvil/compiler/dagger/ProvidesMethodFactoryGeneratorTest.kt]: Nice! Thanks for updating

**Review by JoelWilcox — APPROVED** [review]

**Pull request merged**